[Feature]Added Remove-service to Management module

joandrsn opened the pull request:
I have added the function `Remove-Service` to the Management module.
The functions `New-Service` and `Set-Service` were already present, I thought it would be a good idea to have a `Remove-Service`

Issue #4717

joandrsn committed:
[Feature]Remove-Service added to Management module

joandrsn committed:
[Feature]Capitalization

msftclas commented:
@joandrsn,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,212 @@ protected override void BeginProcessing()
+                    {
+                        hScManager = NativeMethods.OpenSCManagerW(
+                            ServiceComputerName,
+                            null,
```
Please use named parameters.

joandrsn replied:
I don't quite follow you. Did you want to use the variable `ComputerName` ? Since the method supports both a `Name` on the local computer and a servicecontroller on a remote machine, I believe I have to use the `ServiceComputerName` variable.

SteveL-MSFT replied:
see https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments

It makes the code more readable, particularly where null is passed.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -1,4 +1,5 @@
-Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows" {
+Import-Module ".\Microsoft.PowerShell.Commands.Management.dll"
```
This should not be required. Can you remove this?

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -195,6 +196,24 @@ Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows"
+        finally {
+            Get-CimInstance Win32_Service -Filter "name='$servicename'" | Remove-CimInstance -ErrorAction SilentlyContinue
+        }
+    }
```
Can you also add tests which validates the error cases?

adityapatwardhan replied:
Also add tests for pipeline input.

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,212 @@ protected override void BeginProcessing()
+    /// This class implements the set-service command
+    /// </summary>
+    [Cmdlet(VerbsCommon.Remove, "Service", SupportsShouldProcess = true, DefaultParameterSetName = "Name")]
+    [OutputType(typeof(ServiceController))]
```
Typically Remove* cmdlets do not return anything.

joandrsn committed:
[Feature]Added test cases for Remove-Service, Removed return value

markekraus commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,212 @@ protected override void BeginProcessing()
+        #region Parameters
+
+        /// <summary>
+        /// Name of the service to create
```
Should be `Name of the service to remove`?

markekraus commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,212 @@ protected override void BeginProcessing()
+
+        /// <summary>
+        /// The following is the definition of the input parameter "InputObject".
+        /// Specifies ServiceController object representing the services to be stopped.
```
Should be `the services to be removed.`?

markekraus commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,212 @@ protected override void BeginProcessing()
+                            continue;
+                        }
+
+                    
```
Remove extra line.

markekraus reviewed (commented):
Minor formatting and comment/summary rewording suggestions.

joandrsn committed:
[Feature]Documentation from copy-paste code corrected to reflect the new function

joandrsn committed:
[Feature]Erroraction to be sure that an error is thrown in testcase

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -1,4 +1,5 @@
-Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows" {
+Import-Module "C:\Users\jsa\code\PowerShell\src\powershell-win-core\bin\Debug\netcoreapp2.0\win7-x64\Microsoft.PowerShell.Commands.Management.dll"
```
Please Remove

markekraus commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -195,6 +196,47 @@ Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows"
+        { Remove-Service -Name "testremoveservice" -ErrorAction 'Stop' } | Should Throw
+    }
+
+
```
Remove extra line

joandrsn committed:
[Feature]Removed direct reference to module in Pester test

joandrsn committed:
[Feature]Removed extra line in test

joandrsn committed:
[Feature]Use FullyQualifiedErrorId in Remove-Service test without a valid servicename

markekraus commented:
@joandrsn  it looks like the cmdlet doesn't export. I can repro on win10 x64. No clue why, though.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,210 @@ protected override void BeginProcessing()
+        /// <summary>
+        /// Name of the service to remove
+        /// </summary>
+        /// <value></value>
```
should probably remove this

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,210 @@ protected override void BeginProcessing()
+        protected override void ProcessRecord()
+        {
+            ServiceController service = null;
+            string ServiceComputerName = null;
```
for consistency, should use camelCasing

joandrsn commented:
@markekraus If I import the cmdlet with my import statement that I previously committed (and removed again), I can get it to work. Somehow the travis build gets the module imported (it doesn't fail).
I'm not sure why that is. 
If I try a `Get-Command Get-Service` I get a version 3.1.0.0 and after importing the .dll i get 6.0.0.0

adityapatwardhan commented:
@joandrsn I guess you need to add them to the exports here:

https://github.com/PowerShell/PowerShell/blob/master/src/Modules/Windows-Core/Microsoft.PowerShell.Management/Microsoft.PowerShell.Management.psd1

https://github.com/PowerShell/PowerShell/blob/master/src/Modules/Unix/Microsoft.PowerShell.Management/Microsoft.PowerShell.Management.psd1

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,210 @@ protected override void BeginProcessing()
+                                WriteNonTerminatingError(
+                                    service,
+                                    exception,
+                                    "CouldNotRemoveService",
```
This doesn't seem to the right error.  According to the docs, `CloseServiceHandle` only fails if it's an invalid handle which is probably never going to the be case here.  Anyways, even if this fails, it would seem that the service was still deleted.  Maybe this should just be a diagnostic assert?

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,210 @@ protected override void BeginProcessing()
+                            bool succeeded = NativeMethods.CloseServiceHandle(hScManager);
+                            if (!succeeded)
+                            {
+                                int lastError = Marshal.GetLastWin32Error();
```
same as above, perhaps just an assert?

joandrsn committed:
[Feature]Exposed Remove-Service

joandrsn committed:
[Feature]Consistent casing, Named Arguments & Remove incorrect exception on cleanup

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -195,6 +195,46 @@ Describe "Set/New-Service cmdlet tests" -Tags "Feature", "RequireAdminOnWindows"
+            }
+            $service = New-Service @parameters
+            $service | Should Not BeNullOrEmpty
+            Get-Service -Name $servicename | Remove-Service -ErrorAction SilentlyContinue
```
I don't think you want `-ErrorAction SilentlyContinue` there in case something fails

joandrsn committed:
[Feature]Remove-Service test should fail if the service was not removed

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,197 @@ protected override void BeginProcessing()
+                    Diagnostics.Assert(!String.IsNullOrEmpty(Name), "null ServiceName");
+                    // "new ServiceController" will succeed even if
+                    // there is no such service.  This checks whether
+                    // the service actually exists.
```
Should we put the comment before Diagnostics.Assert?
And please reformat the comment  - we can use wider lines.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,197 @@ protected override void BeginProcessing()
 
+    #region RemoveServiceCommand
+    /// <summary>
+    /// This class implements the remove-service command
```
Case - Remove-Service.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,197 @@ protected override void BeginProcessing()
+                bool objServiceShouldBeDisposed = false;
+                try
+                {
+                    if (_ParameterSetName.Equals("InputObject", StringComparison.OrdinalIgnoreCase) && InputObject != null)
```
Why are we using `_ParameterSetName` in the file (twice)? In all other files we use `ParameterSetName`.
I believe we should fix this.

SteveL-MSFT replied:
Looks like line 1604 used this and @joandrsn was just following this pattern.  Agree that we should change it to just `ParameterSetName` (including line 1604) for consistency.

iSazonov replied:
It seems we could remove `_ParameterSetName.Equals("InputObject", StringComparison.OrdinalIgnoreCase)` at all.

SteveL-MSFT replied:
I agree.  InputObject check should be sufficient

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,197 @@ protected override void BeginProcessing()
+                }
+                catch (ArgumentException ex)
+                {
+                    //Cannot use WriteNonterminatingError as service is null
```
Please add space `// Cannot`.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2167,6 +2167,197 @@ protected override void BeginProcessing()
+                }
+                catch (InvalidOperationException ex)
+                {
+                    //Cannot use WriteNonterminatingError as service is null
```
Please add space `// Cannot`

iSazonov commented:
@joandrsn Please add `Remove-Service` in DefaultCommands.Tests.ps1 to pass CI.

joandrsn committed:
[Feature]Cleanup comments & add Remove-Service for CI

markekraus commented on test/powershell/engine/Basic/DefaultCommands.Tests.ps1:
```diff
@@ -407,6 +407,7 @@ Describe "Verify approved aliases list" -Tags "CI" {
 "Cmdlet", "Set-PSDebug",                                        ,                     $($FullCLR -or $CoreWindows -or $CoreUnix)
 "Cmdlet", "Set-PSSessionConfiguration",                         ,                     $($FullCLR -or $CoreWindows -or $CoreUnix)
 "Cmdlet", "Set-Service",                                        ,                     $($FullCLR -or $CoreWindows              )
+"Cmdlet", "Remove-Service",                                     ,                     $($FullCLR -or $CoreWindows              )
```
Can you move this up in the list so it is in alphabetical order? Should go between `Remove-PSSnapin` and `Remove-TypeData`

joandrsn committed:
[Feature]Remove-Service in CI set in alphabetic order

joandrsn committed:
[Feature]Use ParameterSetName instead of _ParameterSetName and rewrite test which used the function class directly

joandrsn commented:
I think I figured it out.
I changed the inheritance from `Cmdlet` to `PSCmdlet` to use `ParameterSetName` instead of `_ParameterSetName`.
This meant I had to rewrite a test which used the class directly. Unfortunatly, the test was too complicated for me to rewrite in the same function as before, so I split the test into 4 separate tests.
Also I'm not sure if the last test actually does anything. On my computer, the service was already running. So maybe I should create a service and then change that service? Looking for feedback.

markekraus commented:
I think this probably needs to be reverted. I believe `ServiceBaseCommand` intentionally derives from `Cmdlet` so that the service cmdlets can be directly invoked. This would likely be a Bucket 1 or 2 Breaking Change.

joandrsn commented:
Do you have a suggestion on how I would be able to use `ParameterSetName` instead? I can't currently see the solution if changing the inheiritance to `PSCmdlet` would be a breaking change.

markekraus commented:
I think it was using `_ParameterSetName` "by design" as `ParameterSetName` is not available in commands that derive from `Cmdlet` instead of `PSCmdlet`. `RemoveServiceCommand` could derive from `PSCmdlet` but that would break the derived methods. `ServiceBaseCommand` could be switched to derive from `PSCmdlet`, but that would probably take a review committee approval.

Looking for guidance from other cmdlets, many of them are using 

```csharp
if (InputObject != null)
```
and
```csharp
if (Name != null)
```

Example [`RenameLocalGroupCommand`](https://github.com/PowerShell/PowerShell/blob/fb7c7b038b720fc78d92662ac3405029b8db625c/src/Microsoft.PowerShell.LocalAccounts/LocalAccounts/Commands/RenameLocalGroupCommand.cs#L188)

I lean towards either switching to the null checks or just leaving it as `_ParameterSetName`. I can see benefits to the Service Cmdlets being directly invoked fro host apps, so Moving to `PSCmdlet` seems like a bad option.

@SteveL-MSFT @iSazonov thoughts?

SteveL-MSFT commented:
Seems like the best solution here (preferring simplicity) is to revert back to deriving from `Cmdlet` as it originally was which perhaps was intentional.  Would like to see the tests back to being `-TestCases` as it seems there a bunch of similar code having split up due to the change in inheritance.

iSazonov commented:
Oops, I again got an artifact 😄 
I unlike using _ParameterSetName because it is marked as "internal" and available through a hack with  `InternalsVisibleTo`.
So I agree that we should revert last changes and remove `_ParameterSetName`.

joandrsn committed:
Revert "[Feature]Use ParameterSetName instead of _ParameterSetName and rewrite test which used the function class directly"

This reverts commit da41f7deb811f32538e0dc7f0d811646ceb8acf8.

joandrsn committed:
[Feature]Remove _ParameterSetName check & added test for pipeline input in
Set-Service

SteveL-MSFT reviewed (approved):
Thanks for the contribution!

adityapatwardhan reviewed (approved):
LGTM

markekraus reviewed (approved):
LGTM

adityapatwardhan commented:
@joandrsn Thanks for your contribution!

PingPongSet commented:
What vesion of PowerShell is Remove-service available for? What about version 4?

PingPongSet commented:
@iSazonov What vesion of PowerShell is Remove-service available for?

iSazonov commented:
@PingPongSet I don' remember the version but you could look release notes for PowerShell Core 6.0. Also make only sense download latest PowerShell Core 6.0 version.
